### PR TITLE
rewrite Cronjob build triggering logic in python; read from influxdb

### PR DESCRIPTION
**Pull request by alsuren**
- [x] rewrite a bunch of the bash build scripts in python
  - [x] dependabot (completely untested)
  - [x] automatic installation of deps when you run relevant `make` rules
  - [x] delete old bash scripts
- [x] switch to reading from influxdb cloud directly, instead of redis via the webserver. See https://github.com/cargo-bins/cargo-quickinstall/pull/165
  - [ ] rebase https://github.com/cargo-bins/cargo-quickinstall/pull/165 and address comments
  - [ ] you now need INFLUXDB_TOKEN to read these stats.
    - [ ] I have added one to our github secrets (with read-only creds)
    - [ ] I need to put everyone in the influx org (different one from the one I already invited everyone too: sorry)
    - [x] I can either make a new public read-only token and put it in the repo (can revoke it if someone eats our rate limits or something) or add a /stats endpoint on the server like the old one had? - you can use `p1YOOpyPWwrcJ5wqeBqpsIVPGMiKDHg_PUkUxFdwl9lW6W5jHLdxL67Fsz4STm1yDd8pwH42YHuX7ZxjRkOu4w==
` for now. I will think about revoking it later.
- [ ] not sure where my print() statements output is disappearing to in the github actions output :(

### CODE REVIEW

File: README.md
Comment (NobodyXu) [inline README.md:48]: Can we keep the `supported-targets`?

cargo-binstall uses it to avoid hitting github APIs, if binstall knows that quickinstall does not support the target.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: Makefile
Comment (NobodyXu) [inline Makefile:8]: I think `$(MAKE) recheck` is better/more canonical way of invoking another makefile?

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [inline cronjob_scripts/trigger-package-build.py:136]:
```suggestion
    rustc_version_output = subprocess.run(
         
["rustc", "--version", "--verbose"], capture_output=True, text=True
    )

    try:
```

Since the `except` access `rustc_version_output`, it can't handle the error from `subprocess.run`

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [inline cronjob_scripts/trigger-package-build.py:211]: Given that we already get all tags from repo, I suppose we can use `git describe --match *`?

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [inline cronjob_scripts/trigger-package-build.py:170]:
```suggestion
                Counter(stripped for line in file if (stripped := line.strip()))
```

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [inline cronjob_scripts/trigger-package-build.py:70]: I think we should remove already built crates, before sampling?

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [inline cronjob_scripts/trigger-package-build.py:72]:
```suggestion
        crates_to_check.append("cargo-quickinstall")
```

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [inline cronjob_scripts/trigger-package-build.py:93]: Nit: could use a generator expression:

```suggestion
                feat for feat in version["features"].keys() if "vendored" in feat
```

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [inline cronjob_scripts/trigger-package-build.py:155]: Something that we could consider using here:

https://github.com/cargo-bins/cargo-quickinstall/issues/268#issuecomment-2329308074

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: cronjob_scripts/get_latest_version.py
Comment (alsuren) [reply cronjob_scripts/get_latest_version.py:18]: needs to work with 3.8 for now

File: cronjob_scripts/checkout_worktree.py
Comment (alsuren) [reply cronjob_scripts/checkout_worktree.py:19]: temp dir won't work here unless I pull the whole thing inside out and turn it into a context manager. For now, I'm just going to keep the logic how it was.

I don't think shutil buys us anything because these scripts are only ever run on unix machines. I'm assuming that if we have maintainers on windows, they all have access to WSL or similar?

In theory I could use `git ls-files $BRANCH` and `git show $BRANCH:$FILE` to avoid having any checkouts of the "trigger/*" branches at all. I don't think this is worth the bother though, because the builders do a checkout when writing failures, and the symmetry is quite nice.

File: cronjob_scripts/get_latest_version.py
Comment (alsuren) [reply cronjob_scripts/get_latest_version.py:34]: annoyingly, the sparse index doesn't calculate this for you. I think my logic is correct, and the sparse index api doesn't have a rate limit, so I'd prefer to use it if possible.

I guess we could grep for max_stable_version in the crates.io source code if we wanted to double-check, but I don't think it's the end of the world if we get it wrong.

Comment (alsuren) [reply cronjob_scripts/get_latest_version.py:34]: in theory, we could make our get_latest_version() in cargo-quickinstall rust client use the crates.io api as well, and keep the logic in sync somehow?

File: cronjob_scripts/trigger-package-build.py
Comment (alsuren) [reply cronjob_scripts/trigger-package-build.py:70]: in order to remove already built crates, we need to hammer the crates.io and github releases apis.

The old code kind-of tried to do this, but then it added this strange 4 * CRATE_CHECK_LIMIT cap to the top **before** shuffling and **before** checking for built packages, so it basically ignores anything that's not in the first ~20 results from get-stats.sh (which I think is sorted alphabetically?).

(go read next-unbuilt-package.sh and you'll see what I mean)

When I run this new code, it finds **loads** of crates that the normal cronjob is failing to pick up, so this is definitely better than before.

I think you have more experience with github's rate limits than me, but I say we ship this algorithm for now and tweak it later.

File: cronjob_scripts/trigger-package-build.py
Comment (alsuren) [reply cronjob_scripts/trigger-package-build.py:72]: it would be .insert(0, "cargo-quickinstall") - I want to check quickinstall first, before the other random crates, because this is used as part of the release process.

File: cronjob_scripts/trigger-package-build.py
Comment (alsuren) [reply cronjob_scripts/trigger-package-build.py:155]: I will add a comment. In practice, that would probably still be put into a text file by some daily cronjob, so the shape of this function probably won't change that much? I haven't really thought about this enough.

File: cronjob_scripts/trigger-package-build.py
Comment (alsuren) [reply cronjob_scripts/trigger-package-build.py:211]: Potentially we could, I guess. This would leave us with the case where we make a tag and then die before we can upload the tarball.

Potentially we could check crates.io + git tags for **all** crates, and pick a small random sample to check the github releases for every hour?

This is the kind of proposal that you just wouldn't dream of doing if the whole thing was written in bash, so I feel quite happy that I made this effort.

I'm not sure whether I want to do this in **this** pr. Maybe a fast follow up?

File: cronjob_scripts/checkout_worktree.py
Comment (NobodyXu) [reply cronjob_scripts/checkout_worktree.py:19]: True, let's keep it for now

File: cronjob_scripts/get_latest_version.py
Comment (NobodyXu) [reply cronjob_scripts/get_latest_version.py:20]: I thought the py env in uv is set to 3.10?

Anyway if the py version is too old, then yeah we should skip this for now

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [reply cronjob_scripts/trigger-package-build.py:122]: crates-io dump is updated daily, and I think we could use `actions/cache` to cache it?

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [reply cronjob_scripts/trigger-package-build.py:59]: Yeah we can always add this later

File: cronjob_scripts/get_latest_version.py
Comment (NobodyXu) [reply cronjob_scripts/get_latest_version.py:20]: Oh I see you downgraded python to 3.8

File: cronjob_scripts/get_latest_version.py
Comment (NobodyXu) [reply cronjob_scripts/get_latest_version.py:36]: Another option would be to use the crates-io dump.

When playing around with it, I discovered that it does have `max_stable_version`

Given that it's updated daily, it should be good enough.

Comment (alsuren) [reply cronjob_scripts/get_latest_version.py:36]: I would like to have so that we can `make release` and have it immediately run `make recheck` and find the latest version of cargo-quickinstall that we just published, and trigger rebuilds for all architectures.

I **think** that the sparse index protocol gives us this, but I've not tested it.

File: cronjob_scripts/trigger-package-build.py
Comment (alsuren) [reply cronjob_scripts/trigger-package-build.py:122]: yeah, possibly, assuming it's not too big. Something for a later PR though.

File: cronjob_scripts/get_latest_version.py
Comment (NobodyXu) [inline cronjob_scripts/get_latest_version.py:61]: Let's optimize this a bit, the crate might have so many versions that it becomes memory intensive, given that popular crates often have a lot of releases as well:

**And I think you didn't handle the situation where there's no available version to build**, while it's unlikely it could happen, so let's handle that as well.

```suggestion
    max_stable_version = None
    for line in response.text.splitlines():
        version = json.loads(line)
        if version["yanked"]:
            continue
        version = semver.VersionInfo.parse(version["vers"])
       if version.prerelease:
           continue
       if max_stable_version is None or version > max_stable_version:
           max_stable_version = version

    return max_stable_version 
```

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [inline cronjob_scripts/trigger-package-build.py:121]: Fixing a bug, where a `"[[:space:]]+#"` is treated as a crate:

```suggestion
            stripped
            for line in file
            if (stripped := line.strip()) and not stripped.startswith("#")
```

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [inline cronjob_scripts/trigger-package-build.py:131]: For these files, I think we don't have any mechanism to delete them?

So it'd keep growing, I'd recon we need to add a cleanup mechanism in another PR, otherwise this would get inefficient eventually.

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [reply cronjob_scripts/trigger-package-build.py:122]: Think it's simple enough, was polars I was able to do it around 10 lines, and it's quite efficient (in the linked issue)

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [reply cronjob_scripts/trigger-package-build.py:178]: > I'm not sure whether I want to do this in this pr. Maybe a fast follow up?

Yeah a follow-up is 👍

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [reply cronjob_scripts/trigger-package-build.py]: cc @alsuren in case you missed this

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [reply cronjob_scripts/trigger-package-build.py]: cc @alsuren in case you missed this

File: cronjob_scripts/get_latest_version.py
Comment (alsuren) [reply cronjob_scripts/get_latest_version.py:61]: I suspect that the RAM usage from parsing the sparse crates.io index will always be dwarfed by the background RAM usage of the python interpreter + libs. Your way does feel a bit cleaner though, and makes the empty result case more obviously correct, so I will refactor it into something more like that shape.

File: cronjob_scripts/trigger-package-build.py
Comment (alsuren) [reply cronjob_scripts/trigger-package-build.py:131]: backlogged as https://github.com/cargo-bins/cargo-quickinstall/issues/276

File: cronjob_scripts/trigger-package-build.py
Comment (alsuren) [reply cronjob_scripts/trigger-package-build.py]: I fixed it when I moved it into architectures.py

File: cronjob_scripts/trigger-package-build.py
Comment (alsuren) [reply cronjob_scripts/trigger-package-build.py]: cheers. I had a bit of clean up of my blanked except: blocks just now, so hopefully it's a bit cleaner now.

File: cronjob_scripts/trigger-package-build.py
Comment (alsuren) [reply cronjob_scripts/trigger-package-build.py:176]: Ah. I thought about this again, and I don't think we can use git tags:

We used to make tags like `cargo-benchcmp-0.4.5-aarch64-apple-darwin` and add a single tarball to it. We now make tags like `cargo-benchcmp-0.4.5` and add a tarball per arch.

This means that we can't just read the output of `git tag` to determine if we have built a thing for a particular arch, because it might have been built for a different architecture.

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [inline cronjob_scripts/trigger-package-build.py:171]: I think using a python library for it would be more readable https://pypi.org/project/PyGithub/

For the rate-limit problem, I think we can provide a read-only GitHub token via secrets.

While `secrets.GITHUB_TOKEN` would work, it
 - has a rate limit of 1000 requests per hour
 - has too many permissions within the repo

If we create a read-only token using a robot account (which I've setup), we will have:
 - 5000 requests per hour
 - has read-only permission and cannot access any private info

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [inline cronjob_scripts/trigger-package-build.py:57]:
```suggestion
    # but I think that shuffling first is an important bugfix. Once we are using crates.io sparse
```

File: cronjob_scripts/architectures.py
Comment (NobodyXu) [inline cronjob_scripts/architectures.py:50]: - strip the line before processing it
 - use `if not` instead of exception handling (I personally don't like exception very much, but if you think it's ok I'd be fine about it).

```suggestion
    current_target = [
        stripped.removeprefix("host: ")
        for line in rustc_version_output.stdout.splitlines()
        if (stripped := line.strip())
        if stripped.startswith("host: ")
    ]
    if not current_target:
        raise ValueError(f"rustc did not tell us its host: {rustc_version_output}")
    return current_target
```

Comment (NobodyXu) [inline cronjob_scripts/architectures.py:50]: - strip the line before processing it
 - use `if not` instead of exception handling (I personally don't like exception very much, but if you think it's ok I'd be fine about it).

```suggestion
    current_target = [
        stripped.removeprefix("host: ")
        for line in rustc_version_output.stdout.splitlines()
        if (stripped := line.strip())
        if stripped.startswith("host: ")
    ]
    if not current_target:
        raise ValueError(f"rustc did not tell us its host: {rustc_version_output}")
    return current_target
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: cronjob_scripts/trigger-package-build.py
Comment (NobodyXu) [reply cronjob_scripts/trigger-package-build.py:171]: @alsuren if you think this is too much work for this PR, we could leave this to a follow-up PR.

Comment (alsuren) [reply cronjob_scripts/trigger-package-build.py:171]: My main goals in this PR are to stop using the old stats server and to delete 400 lines of bash, so I am going to leave it like this.

I also quite like the symmetry of using exactly the same url that the quickinstall client uses. The old code even used the same http client (`curl`) but I decided that the connection pooling from requests was too much of a win.

> We could use the github api to get all artifacts and cache it.

I'm a bit wary of using a "do a lot of work and rely on caches to save us" approach. I've so far been taking the approach of "do as little work as possible and still make forward progress". I realise that there is a cutover though. As we check more crates every hour, we will want to have some kind of cached pre-filter to reduce the number of http requests we need to make. We would need to be wary of stale caches as well: if the cache says the package is there then we can trust that, but if the cache says it's not there, we need to re-check to make sure we didn't build it in the last hour.

In another unrelated note: I think it would be really interesting to gather stats around our build processes as well as our client requests. This would allow us to make better decisions when we make tweaks to the scheduling algorithm. I'm quite annoyed that the new influxdb cloud org doesn't have good notebook/dashboards. I really don't want to have to plug it into grafana just to share some graphs.

**Review by NobodyXu — APPROVED** [review]

**Comment by NobodyXu** [comment]
Thank you, it LGTM, let's merge this in and switch to influx!

i will update the cargo-binstall to use the influx API afterwards and cuts a new release

**Comment by alsuren** [comment]
We also need to get https://github.com/cargo-bins/cargo-quickinstall/pull/165 merged, since that's what's currently powering the influx stuff (note that the vercel stats server is already forwarding stats to the new server, so as long as we get this done in the next month, it should't make much difference).

**Comment by NobodyXu** [comment]
> not sure where my print() statements output is disappearing to in the github actions output :(

For output, we could use grouped output which is more readable

https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#grouping-log-lines

so we can have a 

```python
@contextmanager
def group_output(title):
    print("::group::{title}")
    yield
    print("::endgroup::")

with group_output("Collecting popular crates"):
    print(...)
```

will then looks like image

**Pull request merged**
